drivers/video/video_async.c

```diff
@@ -0,0 +1,182 @@
+struct video_notifier *video_async_find_notifier(const struct device *dev)
+{
+	struct video_notifier *nf;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&notifiers_list, nf, node) {
+		if (nf->dev == dev) {
+			return nf;
+		}
+	}
+
+	return NULL;
+}
+
+/* Find the video device of the notifier tree that this notifier is part of */
+static struct video_device *video_async_find_vdev(struct video_notifier *notifier)
+{
+	if (!notifier) {
+		return NULL;
+	}
+
+	while (notifier->parent) {
+		notifier = notifier->parent;
+	}
+
+	return notifier->vdev;
+}
+
+/* Add all descendants' control handlers of this notifier to the video device */
+static void add_children_ctrl_handlers_to_vdev(struct video_device *vdev, struct video_notifier *nf)
+{
+	if (!vdev || !nf) {
+		return;
+	}
+
+	nf->vdev = vdev;
+
+	video_ctrl_handler_add(&vdev->ctrl_handler, nf->ctrl_handler);
+
+	for (uint8_t i = 0; i < nf->children_num; ++i) {
+		struct video_notifier *nf_child = video_async_find_notifier(nf->children_devs[i]);
+
+		if (!nf_child) {
+			continue;
+		}
+
+		nf_child->vdev = vdev;
+
+		video_ctrl_handler_add(&vdev->ctrl_handler, nf_child->ctrl_handler);
+
+		add_children_ctrl_handlers_to_vdev(vdev, nf_child);
+	}
+}
+
+/* Remove all descendants' control handlers of this notifier from the video device */
+static void remove_children_ctrl_handlers_from_vdev(struct video_device *vdev,
+						    struct video_notifier *nf)
+{
+	if (!vdev || !nf) {
+		return;
+	}
+
+	nf->vdev = NULL;
+
+	video_ctrl_handler_remove(&vdev->ctrl_handler, nf->ctrl_handler);
+
+	for (uint8_t i = 0; i < nf->children_num; ++i) {
+		struct video_notifier *nf_child = video_async_find_notifier(nf->children_devs[i]);
+
+		if (!nf_child) {
+			continue;
+		}
+
+		nf_child->vdev = NULL;
+
+		video_ctrl_handler_remove(&vdev->ctrl_handler, nf_child->ctrl_handler);
+
+		remove_children_ctrl_handlers_from_vdev(vdev, nf_child);
+	}
+}
+
+void video_async_init(struct video_notifier *notifier, const struct device *dev,
+		      struct video_device *vdev, struct video_ctrl_handler *hdl,
+		      const struct device **children_devs, uint8_t children_num)
+{
+	if (!notifier) {
+		return;
+	}
+
+	notifier->dev = dev;
+	notifier->vdev = vdev;
+	notifier->ctrl_handler = hdl;
+	notifier->children_devs = (const struct device **)children_devs;
+	notifier->children_num = children_num;
+}
+
+int video_async_register(struct video_notifier *notifier)
+{
+	uint8_t i = 0;
+	bool found_parent = false;
+	struct video_notifier *nf;
+	struct video_device *vdev = NULL;
+
+	if (!notifier) {
+		return -EINVAL;
+	}
+
+	if (sys_slist_find(&notifiers_list, &notifier->node, NULL)) {
+		LOG_ERR("Notifier of device %s is already registered", notifier->dev->name);
+		return -EALREADY;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&notifiers_list, nf, node) {
+		/* Find the parent notifier in the global notifiers registry */
+		if (!found_parent) {
+			for (i = 0; i < nf->children_num; ++i) {
+				if (nf->children_devs[i] == notifier->dev) {
```

Comment:
Trying to reduce nested ifs :)
```suggestion
				if (nf->children_devs[i] != notifier->dev) {
				    continue;
				}
```